polkadot/node/tracking-allocator/src/lib.rs

```diff
@@ -0,0 +1,171 @@
+	sync::atomic::{AtomicBool, Ordering},
+};
+
+struct TrackingAllocatorData {
+	lock: AtomicBool,
+	current: isize,
+	peak: isize,
+	limit: isize,
+	failure_handler: Option<Box<dyn Fn()>>,
+}
+
+impl TrackingAllocatorData {
+	#[inline]
+	fn lock(&self) {
+		loop {
+			// Try to acquire the lock.
+			if self
+				.lock
+				.compare_exchange_weak(false, true, Ordering::Acquire, Ordering::Relaxed)
+				.is_ok()
+			{
+				break
+			}
+			// We failed to acquire the lock; wait until it's unlocked.
+			//
+			// In theory this should result in less coherency traffic as unlike `compare_exchange`
+			// it is a read-only operation, so multiple cores can execute it simultaneously
+			// without taking an exclusive lock over the cache line.
+			while self.lock.load(Ordering::Relaxed) {
+				std::hint::spin_loop();
+			}
+		}
+	}
+
+	#[inline]
+	fn unlock(&self) {
+		self.lock.store(false, Ordering::Release);
+	}
+
+	fn start_tracking(&mut self, limit: isize, failure_handler: Option<Box<dyn Fn()>>) {
+		self.lock();
+		self.current = 0;
+		self.peak = 0;
+		self.limit = limit;
+		self.failure_handler = failure_handler;
+		self.unlock();
+	}
+
+	fn end_tracking(&mut self) -> isize {
+		self.lock();
+		let peak = self.peak;
+		self.limit = 0;
+		self.failure_handler = None;
+		self.unlock();
+		peak
+	}
+
+	#[inline]
+	fn track(&mut self, alloc: isize) -> bool {
+		self.lock();
+		self.current += alloc;
+		if self.current > self.peak {
+			self.peak = self.current;
+		}
+		let within_limits = self.limit == 0 || self.peak <= self.limit;
+		if within_limits {
+			self.unlock()
+		}
+		within_limits
+	}
+}
+
+static mut ALLOCATOR_DATA: TrackingAllocatorData = TrackingAllocatorData {
+	lock: AtomicBool::new(false),
+	current: 0,
+	peak: 0,
+	limit: 0,
+	failure_handler: None,
+};
+
+pub struct TrackingAllocator<A: GlobalAlloc>(pub A);
+
+impl<A: GlobalAlloc> TrackingAllocator<A> {
+	// SAFETY:
+	// * The following functions write to `static mut`. That is safe as the critical section inside
+	//   is isolated by an exclusive lock.
+
+	/// Start tracking
+	pub fn start_tracking(&self, limit: Option<isize>, failure_handler: Option<Box<dyn Fn()>>) {
+		unsafe {
+			ALLOCATOR_DATA.start_tracking(limit.unwrap_or(0), failure_handler);
+		}
+	}
+
+	/// End tracking and return the peak allocation value in bytes (as `isize`). Peak allocation
+	/// value is not guaranteed to be neither non-zero nor positive.
+	pub fn end_tracking(&self) -> isize {
+		unsafe { ALLOCATOR_DATA.end_tracking() }
+	}
+}
+
+unsafe impl<A: GlobalAlloc> GlobalAlloc for TrackingAllocator<A> {
+	// SAFETY:
+	// * The wrapped methods are as safe as the underlying allocator implementation is
+
+	#[inline]
+	unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+		if ALLOCATOR_DATA.track(layout.size() as isize) {
+			self.0.alloc(layout)
+		} else {
+			if let Some(failure_handler) = &ALLOCATOR_DATA.failure_handler {
+				failure_handler()
+			}
+			ALLOCATOR_DATA.unlock();
+			null_mut()
```

Comment:
Move this into a separate function and mark it `#[cold]` so that the compiler can generate more efficient code.

You can have the function return `*mut u8`, that way the compiler should be able to directly `jmp` to it.